// @dendronhq/common-all/lib/types/workspace.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.DuplicateNoteAction = exports.NoteAddBehavior = exports.DVaultSync = exports.DVaultVisibility = void 0;
var DVaultVisibility;
(function (DVaultVisibility) {
    DVaultVisibility["PRIVATE"] = "private";
})(DVaultVisibility = exports.DVaultVisibility || (exports.DVaultVisibility = {}));
var DVaultSync;
(function (DVaultSync) {
    DVaultSync["SKIP"] = "skip";
    DVaultSync["NO_PUSH"] = "noPush";
    DVaultSync["NO_COMMIT"] = "noCommit";
    DVaultSync["SYNC"] = "sync";
})(DVaultSync = exports.DVaultSync || (exports.DVaultSync = {}));
var NoteAddBehavior;
(function (NoteAddBehavior) {
    NoteAddBehavior["childOfDomain"] = "childOfDomain";
    NoteAddBehavior["childOfDomainNamespace"] = "childOfDomainNamespace";
    NoteAddBehavior["childOfCurrent"] = "childOfCurrent";
    NoteAddBehavior["asOwnDomain"] = "asOwnDomain";
})(NoteAddBehavior = exports.NoteAddBehavior || (exports.NoteAddBehavior = {}));
var DuplicateNoteAction;
(function (DuplicateNoteAction) {
    DuplicateNoteAction["USE_VAULT"] = "useVault";
})(DuplicateNoteAction = exports.DuplicateNoteAction || (exports.DuplicateNoteAction = {}));
