// @dendronhq/engine-server/lib/seed/registry.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.SeedRegistry = void 0;
const common_server_1 = require("@dendronhq/common-server");
const SEED_REGISTRY = {
    "dendron.dendron-site": {
        name: "dendron-site",
        publisher: "dendron",
        description: "Dendron site docs",
        license: "Creative Commons Attribution 4.0 International",
        root: "vault",
        repository: {
            type: "git",
            url: "git@github.com:dendronhq/dendron-site.git",
        },
        site: {
            url: "https://wiki.dendron.so",
            index: "dendron",
        },
    },
    "dendron.handbook": {
        name: "handbook",
        publisher: "dendron",
        description: "Dendron Public Handbook",
        license: "Creative Commons Attribution 4.0 International",
        root: "handbook",
        repository: {
            type: "git",
            url: "git@github.com:dendronhq/handbook.git",
        },
        site: {
            url: "https://handbook.dendron.so",
            index: "handbook",
        },
    },
    "dendron.tldr": {
        name: "tldr",
        publisher: "dendron",
        description: "CLI Docs",
        license: "Creative Commons Attribution 4.0 International",
        root: "vault",
        repository: {
            type: "git",
            url: "https://github.com/kevinslin/seed-tldr.git",
        },
        site: {
            url: "https://tldr.dendron.so",
            index: "cli",
        },
    },
};
class SeedRegistry {
    constructor(registry) {
        this.registry = registry;
    }
    static create(opts) {
        let registry = SEED_REGISTRY;
        if (opts === null || opts === void 0 ? void 0 : opts.registryFile) {
            registry = common_server_1.readYAML(opts.registryFile);
        }
        return new SeedRegistry(registry);
    }
    info({ id }) {
        return this.registry[id];
    }
}
exports.SeedRegistry = SeedRegistry;
