// @dendronhq/dendron-cli/lib/src/commands/launchEngineServer.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.LaunchEngineServerCommand = void 0;
const api_server_1 = require("@dendronhq/api-server");
const common_server_1 = require("@dendronhq/common-server");
const engine_server_1 = require("@dendronhq/engine-server");
const lodash_1 = __importDefault(require("lodash"));
const base_1 = require("./base");
class LaunchEngineServerCommand extends base_1.CLICommand {
    constructor() {
        super({
            name: "launchEngineServer",
            desc: "launch instance of dendron engine",
        });
    }
    buildArgs(args) {
        super.buildArgs(args);
        args.option("port", {
            describe: "port to launch server",
            type: "number",
        });
        args.option("init", {
            describe: "initialize server",
            type: "boolean",
        });
        args.option("noWritePort", {
            describe: "don't write the port to a file",
            type: "boolean",
        });
    }
    async enrichArgs(args) {
        const ctx = "enrichArgs";
        let { wsRoot, port, init, noWritePort } = lodash_1.default.defaults(args, {
            init: false,
            noWritePort: false,
        });
        wsRoot = common_server_1.resolvePath(wsRoot, process.cwd());
        const ws = new engine_server_1.WorkspaceService({ wsRoot });
        const { vaults, dev } = ws.config;
        const vaultPaths = vaults.map((v) => common_server_1.resolvePath(v.fsPath, wsRoot));
        const { port: _port, server } = await api_server_1.launchv2({
            port,
            logPath: process.env["LOG_DST"],
            logLevel: process.env["LOG_LEVEL"] || "error",
            nextServerUrl: dev === null || dev === void 0 ? void 0 : dev.nextServerUrl,
            nextStaticRoot: dev === null || dev === void 0 ? void 0 : dev.nextStaticRoot,
        });
        ws.writeMeta({ version: "dendron-cli" });
        if (!noWritePort) {
            ws.writePort(_port);
        }
        const engine = engine_server_1.DendronEngineClient.create({
            port: _port,
            vaults,
            ws: wsRoot,
        });
        if (init) {
            this.L.info({ ctx, msg: "pre:engine.init" });
            await engine.init();
        }
        return {
            ...args,
            engine,
            wsRoot,
            init,
            vaults: vaultPaths,
            port: _port,
            server,
        };
    }
    async execute(opts) {
        const { port, server } = opts;
        return {
            port: lodash_1.default.toInteger(port),
            server,
        };
    }
}
exports.LaunchEngineServerCommand = LaunchEngineServerCommand;
