// @dendronhq/dendron-cli/lib/src/commands/publishPod.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.PublishPodCLICommand = void 0;
const pods_core_1 = require("@dendronhq/pods-core");
const base_1 = require("./base");
const pod_1 = require("./pod");
const utils_1 = require("./utils");
class PublishPodCLICommand extends base_1.CLICommand {
    constructor() {
        super({
            name: "publishPod",
            desc: "publish a note",
        });
    }
    buildArgs(args) {
        super.buildArgs(args);
        utils_1.setupEngineArgs(args);
        pod_1.setupPodArgs(args);
    }
    async enrichArgs(args) {
        return pod_1.enrichPodArgs({ pods: pods_core_1.getAllPublishPods(), podType: "publish" })(args);
    }
    async execute(opts) {
        const { podClass: PodClass, config, wsRoot, engine, server } = opts;
        const vaults = engine.vaults;
        const pod = new PodClass();
        const resp = await pod.execute({ wsRoot, config, engine, vaults });
        if (config.dest === "stdout") {
            console.log(resp);
        }
        server.close((err) => {
            if (err) {
                this.L.error({ msg: "error closing", payload: err });
            }
        });
    }
}
exports.PublishPodCLICommand = PublishPodCLICommand;
