// @dendronhq/dendron-cli/lib/src/commands/workspaceCLICommand.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.WorkspaceCLICommand = exports.WorkspaceCommands = void 0;
const engine_server_1 = require("@dendronhq/engine-server");
const base_1 = require("./base");
const utils_1 = require("./utils");
var WorkspaceCommands;
(function (WorkspaceCommands) {
    WorkspaceCommands["PULL"] = "pull";
    WorkspaceCommands["PUSH"] = "push";
    WorkspaceCommands["ADD_AND_COMMIT"] = "addAndCommit";
    WorkspaceCommands["SYNC"] = "sync";
    WorkspaceCommands["REMOVE_CACHE"] = "removeCache";
})(WorkspaceCommands = exports.WorkspaceCommands || (exports.WorkspaceCommands = {}));
class WorkspaceCLICommand extends base_1.CLICommand {
    constructor() {
        super({ name: "workspace <cmd>", desc: "workspace related commands" });
    }
    buildArgs(args) {
        super.buildArgs(args);
        utils_1.setupEngineArgs(args);
        args.positional("cmd", {
            describe: "a command to run",
            choices: Object.values(WorkspaceCommands),
            type: "string",
        });
    }
    async enrichArgs(args) {
        const engineOpts = { ...args, init: false };
        const engineArgs = await utils_1.setupEngine(engineOpts);
        return { ...args, ...engineArgs };
    }
    async execute(opts) {
        const { cmd, wsRoot } = opts;
        try {
            switch (cmd) {
                case WorkspaceCommands.PULL: {
                    const ws = new engine_server_1.WorkspaceService({ wsRoot });
                    await ws.pullVaults();
                    break;
                }
                case WorkspaceCommands.ADD_AND_COMMIT: {
                    const ws = new engine_server_1.WorkspaceService({ wsRoot });
                    await ws.commitAndAddAll();
                    break;
                }
                case WorkspaceCommands.PUSH: {
                    const ws = new engine_server_1.WorkspaceService({ wsRoot });
                    await ws.pushVaults();
                    break;
                }
                case WorkspaceCommands.REMOVE_CACHE: {
                    const ws = new engine_server_1.WorkspaceService({ wsRoot });
                    await ws.removeVaultCaches();
                    this.print("caches removed");
                    break;
                }
                case WorkspaceCommands.SYNC: {
                    const ws = new engine_server_1.WorkspaceService({ wsRoot });
                    this.print("commit and add...");
                    await ws.commitAndAddAll();
                    this.print("pull...");
                    await ws.pullVaults();
                    this.print("push...");
                    await ws.pushVaults();
                    this.print("done...");
                    break;
                }
                default: {
                    throw Error("bad option");
                }
            }
        }
        catch (err) {
            this.L.error(err);
        }
        finally {
            if (opts.server.close) {
                opts.server.close();
            }
        }
    }
}
exports.WorkspaceCLICommand = WorkspaceCLICommand;
