// @dendronhq/pods-core/lib/builtin/HTMLPod.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.HTMLPublishPod = void 0;
const engine_server_1 = require("@dendronhq/engine-server");
const basev3_1 = require("../basev3");
const ID = "dendron.html";
class HTMLPublishPod extends basev3_1.PublishPod {
    async plant(opts) {
        const { config, engine, note } = opts;
        const { fname } = config;
        const { data: econfig } = await engine.getConfig();
        const proc = engine_server_1.MDUtilsV4.procFull({
            engine,
            dest: engine_server_1.DendronASTDest.HTML,
            vault: note.vault,
            fname,
            shouldApplyPublishRules: false,
            publishOpts: {
                insertTitle: econfig.useFMTitle,
            },
            config: econfig,
            mermaid: econfig.mermaid,
        });
        const { contents } = await engine_server_1.MDUtilsV4.procRehype({
            proc,
            mathjax: true,
        }).process(note.body);
        return contents;
    }
}
exports.HTMLPublishPod = HTMLPublishPod;
HTMLPublishPod.id = ID;
HTMLPublishPod.description = "publish html";
